core/src/test/java/org/elasticsearch/index/translog/MultiSnapshotTests.java

```diff
@@ -0,0 +1,82 @@
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.translog;
+
+import com.carrotsearch.hppc.LongHashSet;
+import com.carrotsearch.hppc.LongSet;
+import org.elasticsearch.common.Randomness;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.LongStream;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+
+public class MultiSnapshotTests extends ESTestCase {
+    public void testTrackSeqNumRandomRanges() throws Exception {
+        final MultiSnapshot.SeqNumSet bitSet = new MultiSnapshot.SeqNumSet();
+        final LongSet normalSet = new LongHashSet();
+        IntStream.range(0, between(20_000, 50_000)).forEach(i -> {
+            long seq = randomNonNegativeLong();
+            boolean existed = normalSet.add(seq) == false;
+            assertThat("SeqNumSet != Set", bitSet.getAndSet(seq), equalTo(existed));
+        });
+    }
+
+    public void testTrackSeqNumDenseRanges() throws Exception {
+        final MultiSnapshot.SeqNumSet bitSet = new MultiSnapshot.SeqNumSet();
+        final LongSet normalSet = new LongHashSet();
+        IntStream.range(0, between(20_000, 50_000)).forEach(i -> {
+            long seq = between(0, 5000);
+            boolean existed = normalSet.add(seq) == false;
+            assertThat("SeqNumSet != Set" + seq, bitSet.getAndSet(seq), equalTo(existed));
+        });
+    }
+
+    public void testTrackSeqNumSparseRanges() throws Exception {
+        final MultiSnapshot.SeqNumSet bitSet = new MultiSnapshot.SeqNumSet();
+        final LongSet normalSet = new LongHashSet();
+        IntStream.range(0, between(20_000, 50_000)).forEach(i -> {
+            long seq = between(i * 10_000, i * 30_000);
+            boolean existed = normalSet.add(seq) == false;
+            assertThat("SeqNumSet != Set", bitSet.getAndSet(seq), equalTo(existed));
+        });
+    }
+
+    public void testSequenceNumMimicTranslog() throws Exception {
+        final MultiSnapshot.SeqNumSet bitSet = new MultiSnapshot.SeqNumSet();
+        final LongSet normalSet = new LongHashSet();
+        long currentSeq = between(10_000_000, 1_000_000_000);
+        final int iterations = between(100, 2000);
+        for (long i = 0; i < iterations; i++) {
+            List<Long> batch = LongStream.range(currentSeq, currentSeq + between(1, 1000))
```

Comment:
can we make sure batches sometimes crossed the underlying 1024 bitset implementation, to check that we handle completed sets correctly?  Maybe we can also expose the number of completed sets in SeqNumSet and test that we see the right number.

Reply:
:+1: I'd also quite like to see some non-random tests that _deliberately_ hit the corners, rather than relying on the random ones finding everything, and also think it'd be useful to assert that there are the expected number of complete/incomplete sets.

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/27268/commits/ea11a2adc28b6c5b8b6498df8efb1661698f7c3f